File: Data Structures & Algorithms/trapping-rain-water/submission-2.cpp
class Solution {
   public:
    int trap(vector<int>& height) {
        int result = 0;
        vector<int> prefix;
        vector<int> suffix;
        int prefixMax = 0;
        int suffixMax = 0;

        for (int i = 0; i < height.size(); ++i) {
            prefixMax = max(prefixMax, height[i]);
            suffixMax = max(suffixMax, height[height.size() - 1 - i]);
            prefix.push_back(prefixMax);
            suffix.push_back(suffixMax);
        }

        for (int i = 0; i < height.size(); ++i) {
            result += min(prefix[i], suffix[height.size() - 1 - i]) - height[i];
        }

        return result;
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        nums_size = len(nums)
        result = []
        prefix, suffix = 1, 1

        for i in range(nums_size):
            result.append(prefix)
            prefix *= nums[i]

        for i in range(nums_size - 1, -1, -1):
            result[i] *= suffix
            suffix *= nums[i]

        return result
